File: Racing.Moto.Web/Areas/Admin/Controllers/UserController.cs
using NLog;
using Racing.Moto.Core.Extentions;
using Racing.Moto.Data.Entities;
using Racing.Moto.Data.Enums;
using Racing.Moto.Data.Models;
using Racing.Moto.Services;
using Racing.Moto.Services.Constants;
using Racing.Moto.Services.Mvc;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace Racing.Moto.Web.Areas.Admin.Controllers
{
    [Authorize]
    public class UserController : BaseController
    {
        private ILogger _logger = LogManager.GetCurrentClassLogger();

        #region 设计版

        #region 所有用户

        public ActionResult All()
        {
            ViewBag.UserType = UserType.All;

            return View();
        }

        #endregion

        #region 游客

        public ActionResult Vistor()
        {
            //ViewBag.UserType = UserType.Vistor;

            return View();
        }

        #endregion

        #endregion

        public JsonResult GetUsers(UserSearchModel searchModel)
        {
            var result = new ResponseResult();

            try
            {
                var userService = new UserService();

                if (userService.SearchEnabled(LoginUser.UserId, searchModel))
                {
                    var pager = new UserService().GetUsers(searchModel);

                    // 在线用户
                    var onlienUsers = PKBag.OnlineUserRecorder.GetUserList();
                    foreach (var user in pager.Items)
                    {
                        user.IsOnline = onlienUsers.Where(u => u.UserName == user.UserName).Any();
                    }

                    result.Data = pager;
                }
                else
                {
                    result.Data = new PagerResult<User>();
                }
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }

        #region 网站参考版

        #region 用户管理

        #region 总代理

        public ActionResult GeneralAgent()
        {
            ViewBag.UserType = UserType.GeneralAgent;

            return View();
        }

        // 添加 or 修改
        public ActionResult GeneralAgentManagement(int id = 0)
        {
            ViewBag.UserType = UserType.GeneralAgent;
            ViewBag.UserId = id; // =0: add, >0 edit

            return View();
        }

        #endregion

        #region 代理

        /// <summary>
        /// 代理
        /// </summary>
        /// <param name="id">上级UserID</param>
        /// <returns></returns>
        public ActionResult Agent(int id = 0)
        {
            ViewBag.UserType = UserType.Agent;
            ViewBag.FatherUserId = id;

            return View();
        }

        // 添加 or 修改
        public ActionResult AgentManagement(int id = 0)
        {
            ViewBag.UserType = UserType.Agent;
            ViewBag.UserId = id; // =0: add, >0 edit

            return View();
        }

        #endregion

        #region 会员

        public ActionResult Member(int id = 0, int cid = 0)
        {
            ViewBag.UserType = UserType.Member;
            ViewBag.FatherUserId = id;
            ViewBag.GrandfatherUserId = cid;

            return View();
        }

        // 添加 or 修改
        public ActionResult MemberManagement(int id = 0)
        {
            ViewBag.UserType = UserType.Member;
            ViewBag.UserId = id; // =0: add, >0 edit

            return View();
        }

        #endregion

        ///// <summary>
        ///// 用户
        ///// </summary>
        ///// <param name="id">2:总代理, 3: 代理, 4: 会员</param>
        ///// <returns></returns>
        //public ActionResult List(int id)
        //{
        //    ViewBag.UserType = id;

        //    return View();
        //}

        [HttpPost]
        public ActionResult GetUser(int id)
        {
            var result = new ResponseResult();

            try
            {
                result.Data = new UserService().GetUser(id);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }

        [HttpPost]
        public JsonResult SaveUser(int type, User user)
        {
            var result = new ResponseResult();

            try
            {
                result = new UserService().SaveUser(type, user);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }

        /// <summary>
        /// 删除
        /// </summary>
        /// <param name="id">userId</param>
        /// <param name="enabled">是否可用</param>
        [HttpPost]
        public ActionResult RemoveUser(int id, bool enabled)
        {
            var result = new ResponseResult();

            try
            {
                new UserService().EnableUser(id, enabled);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }


        /// <summary>
        /// 冻结
        /// </summary>
        /// <param name="id">userId</param>
        /// <param name="locked">是否冻结</param>
        [HttpPost]
        public ActionResult LockUser(int id, bool locked)
        {
            var result = new ResponseResult();

            try
            {
                new UserService().LockUser(id, locked);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }



        #region 退水

        public ActionResult Rebate(int id, int cid = 0)
        {
            ViewBag.UserType = id;
            ViewBag.UserId = cid;
            ViewBag.UserTypeName = RoleConst.GetRoleName(id);
            ViewBag.RebateUser = new UserService().GetUserOnly(cid);

            return View();
        }

        [HttpPost]
        public JsonResult GetRebates(int id)
        {
            var result = new ResponseResult();

            try
            {
                result.Data = new UserRebateService().GetUserRebates(id);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }

        [HttpPost]
        public JsonResult SaveRebates(int userId, List<UserRebate> userRebates)
        {
            var result = new ResponseResult();

            try
            {
                new UserRebateService().UpdateUserRebates(userId, userRebates);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.Message = MessageConst.System_Error;
                _logger.Info(ex);
            }

            return Json(result);
        }

        #endregion

        #endregion

        #endregion
    }
}

File: Racing.Moto.Web.Admin/Scripts/Controllers/Admin/UserController.js
app.controller('userController', ['$scope', '$rootScope', '$http', '$compile', '$timeout', '$q', '$sce', function ($scope, $rootScope, $http, $compile, $timeout, $q, $sce) {
    $scope.data = {
        UserTypes: { All: 0, Admin: 1, GeneralAgent: 2, Agent: 3, Member: 4, Vistor: 5 },
    };
    $scope.init = function (userType) {
        $scope.user.init(userType);
    };

    $scope.user = {
        UserType: null,
        IsEdit: false,
        CurrentUser: null,
        init: function (userType) {
            $scope.user.UserType = userType;
            $scope.pager.init(userType);
        },
        edit: function () {
            $scope.user.IsEdit = true;
        },
        save: function () {
            var data = {
                type: $scope.user.UserType,
                users: $scope.user.Data
            };
            $http.post('/PK/SaveUsers', data).then(function (res) {
                console.log(res);
                if (res.data.Success) {
                    $scope.user.DataBak = angular.copy($scope.user.Data);
                    $scope.user.IsEdit = false;

                    alert('修改成功!');
                } else {
                    alert(res.data.Message);
                }
            });
        },
        revert: function () {
            $scope.user.Data = angular.copy($scope.user.DataBak);
        },
    };

    $scope.pager = {
        PageIndex: 1,
        PageSize: 15,
        RowCount: 0,
        Params: {
            UserType: $scope.data.UserTypes.All,
            PageIndex: 1,
            PageSize: 15
        },
        Results: [],
        init: function (userType) {
            $scope.pager.Params.UserType = userType;
            $scope.pager.getResults(1);
        },
        getResults: function (pageIndex) {
            $scope.pager.Params.PageIndex = pageIndex;
            $http.post('/User/GetUsers', $scope.pager.Params).then(function (res) {
                console.log(res);
                if (res.data.Success) {
                    $scope.pager.Results = res.data.Data.Items;
                    $scope.pager.RowCount = res.data.Data.RowCount;
                } else {
                    alert(res.data.Message)
                }
            });
        },
    };
}]);

//网站参考版
app.controller('userListController', ['$scope', '$rootScope', '$http', '$compile', '$timeout', '$q', '$sce', function ($scope, $rootScope, $http, $compile, $timeout, $q, $sce) {
    $scope.data = {
        UserTypes: { All: 0, Admin: 1, GeneralAgent: 2, Agent: 3, Member: 4, Vistor: 5 },
    };

    $scope.init = function (userType, fatherUserId, grandFatherUserId) {
        $scope.user.init(userType, fatherUserId, grandFatherUserId);
    };

    $scope.user = {
        UserType: null,
        FatherUserId: null,
        GrandFatherUserId: null,
        init: function (userType, fatherUserId, grandFatherUserId) {
            $scope.user.UserType = parseInt(userType, 10);
            $scope.user.FatherUserId = fatherUserId;
            $scope.user.GrandFatherUserId = grandFatherUserId;
            $scope.pager.init(userType, fatherUserId, grandFatherUserId);
        },
        addUser: function (userType) {
            location.href = $scope.user.getUrl();
        },
        editUser: function (userId) {
            location.href = $scope.user.getUrl() + '/' + userId;
        },
        rebate: function (userId) {
            location.href = '/user/rebate/' + $scope.user.UserType + '/' + userId;
        },
        getUrl: function () {
            var url = '';

            switch ($scope.user.UserType) {
                case $scope.data.UserTypes.GeneralAgent: url = '/user/GeneralAgentManagement'; break;
                case $scope.data.UserTypes.Agent: url = '/user/AgentManagement'; break;
                case $scope.data.UserTypes.Member: url = '/user/MemberManagement'; break;
            }

            return url;
        },
        removeUser: function (userId, enabled) {
            if (confirm('确定删除?')) {
                $http.post('/User/RemoveUser', { id: userId, enabled: enabled }).then(function (res) {
                    console.log(res);
                    if (res.data.Success) {
                        alert('操作成功!');
                        $scope.pager.getResults($scope.pager.PageIndex);
                    } else {
                        alert(res.data.Message)
                    }
                });
            }
        },
        lockUser: function (userId, locked) {
            if (confirm('确定冻结?')) {
                $http.post('/User/LockUser', { id: userId, locked: locked }).then(function (res) {
                    console.log(res);
                    if (res.data.Success) {
                        alert('操作成功!');
                        $scope.pager.getResults($scope.pager.PageIndex);
                    } else {
                        alert(res.data.Message)
                    }
                });
            }
        },
    };

    $scope.pager = {
        PageIndex: 1,
        PageSize: 15,
        RowCount: 0,
        Params: {
            FatherUserId: 0,
            GrandFatherUserId: 0,
            UserType: $scope.data.UserTypes.Agent,
            //IsLocked: 'false',
            PageIndex: 1,
            PageSize: 15
        },
        Results: [],
        init: function (userType, fatherUserId, grandFatherUserId) {
            $scope.pager.Params.UserType = userType;
            $scope.pager.Params.FatherUserId = fatherUserId;
            $scope.pager.Params.GrandFatherUserId = grandFatherUserId;
            $scope.pager.getResults(1);
        },
        getResults: function (pageIndex) {
            $scope.pager.Params.PageIndex = pageIndex;
            $http.post('/User/GetUsers', $scope.pager.Params).then(function (res) {
                console.log(res);
                if (res.data.Success) {
                    $scope.pager.Results = res.data.Data.Items;
                    $scope.pager.RowCount = res.data.Data.RowCount;
                } else {
                    alert(res.data.Message)
                }
            });
        },
        pageChanged: function () {
            $scope.pager.getResults($scope.pager.PageIndex);
        },
    };
}]);